### ISSUE TITLE
select: provide a way for authors to define the string value of rich options

### ISSUE THREAD

**Issue by smhigley** [openui/open-ui#1118]
Introducing the ability to render structured rich content within options also introduces the need for authors to be able to separately define the rendered DOM in the option vs. the string-typed text used for features like typeahead / type-to-navigate.

I ran into this exact issue while creating the Dropdown & Combobox controls in the library I work on. A concrete example can be found in UI like this:
<img width="278" alt="Screenshot of an open select-like react component with the visible text value Elvia Atkins displayed in the trigger button. In the open menu are four options. Each option has a circular avatar graphic, a bolded name, and smaller text underneath the name showing a status like busy, available, and out of office." src="https://github.com/user-attachments/assets/3d90fd9e-8b79-4970-82c1-3f6fb78702cd">

When displaying the selected value in the trigger button we only want to display the text of the person's name. This is currently possible using `<selectedoption>` + custom CSS `display: none` styles. What isn't currently possible is enabling users to type e.g "Elv" to jump to "Elvia Atkins", Or repeatedly typing the same letter to cycle through multiple people whose names all start with that letter.

The best we can do as a browser fallback is to use the text content of the entire option, which is not always conceptually the text that a user would try to type. That still seems fine to me as a default, but authors who want to do the right thing currently have no ability to customize this.

The other potential reason for needing this is if we keep the same `<option>` model & attributes in a future editable select, there will need to be a way to define what string-only value gets inserted into the `<input>` when an option is selected.

## Solutions

There are a couple current HTML attributes on the option element that are at least conceptually similar to this idea:

- `label`: right now this overrides the displayed text of the `<option>` if defined. Repurposing this to no longer override the text could be considered breaking (maybe?), although otherwise this feels like a good fit.
- `value`: even though the word "value" captures the idea of this feature, repurposing current behavior of `value` -- which is the form-submitted non-human-readable value -- would be a bigger and more breaking change.

The other option (heh) is to create a new HTML attribute specifically for the string version of option's primary text. This is what we do in my library, using `text`. For example, the options in the above screenshot have roughly this code:

```html
<Option text="Elvia Atkins" value="eatkins">
  <Avatar />
  Elvia Atkins
  <span class="status">Busy</span>
</Option>
```

## Open questions:
- Assuming we add or repurpose an attribute for this, should that attribute ever be used as the displayed text in the select trigger?
- Should it ever be used in the calculation of the accessible value string? (which should probably match the displayed text most of the time)

Related: #1117

**Comment by josepharhar** [comment]
> `label`: right now this overrides the displayed text of the `<option>` if defined. Repurposing this to no longer override the text could be considered breaking (maybe?), although otherwise this feels like a good fit.

I agree it feels like a good fit, but going through with it is probably going to be complicated. The current behavior is very unfortunate. Hopefully we can just stop rendering the label attribute for the new mode: https://github.com/openui/open-ui/issues/1115

> `value`: even though the word "value" captures the idea of this feature, repurposing current behavior of `value` -- which is the form-submitted non-human-readable value -- would be a bigger and more breaking change.

Yeah I'm not sure the value attribute should be used for typeahead or anything thats exposed to the user at all.

> Assuming we add or repurpose an attribute for this, should that attribute ever be used as the displayed text in the select trigger?

It sounds like ideally we would not use the label attribute when rendering anything so that it could be used for this use case.

> Should it ever be used in the calculation of the accessible value string? (which should probably match the displayed text most of the time)

Great question!

**Comment by josepharhar** [comment]
I think we still need to figure out how we are going to incorproate the option label attribute for accessibility mappings, but

Proposed resolution: Use the option's label attribute for typeahead. If there is no label attribute, then use the option's text content for typeahead.

**Comment by josepharhar** [comment]
The tone about standardizing keyboard behavior makes me believe that getting interoperable typeahead isn't going to be possible: https://github.com/openui/open-ui/issues/1087

I think this means that we can't add a new attribute for typeahead, but I'm happy to implement whatever typeahead behavior you think is best in chromium.

**Comment by gregwhitworth** [comment]
While not a resolution - an important action was made for @gregwhitworth and @smhigley here

> <jarhar> sarah: greg would you be open to working with me outside of this? i would like to get a solution quickly. swiss army knife analogy, i would think of this more like a blade lock rather than an additional feature just to make the new stuff more safe
> <jarhar> gregwhitworth: i can bring in the person who helped write that from internal meetings. we can spread it across 3 people
> <jarhar> masonf: i think it would be great to pull it out in a separate explainer

**Comment by josepharhar** [comment]
A similar topic came up for comobox (and filterable select?) where you might want to specify the text to match filtering apis against for each option element, such as alternate spellings of countries in a country picker.

If you're implementing your own filtering then maybe it wouldn't be necessary, but for builtin filtering it could be useful.

**Comment by LeaVerou** [comment]
`label` is the obvious choice. This is not only useful for the use cases in the OP, but also for specifying a fallback for non-customizable_select browsers that differs from the automatic textContent of each option. This is currently incredibly annoying to do for some cases, and basically requires duplicating the option for some of the more complex ones.

I've even seen agents using it that way, then realizing it didn't work. And it's a relatively unknown attribute, with little existing purpose. Customizable select is used very little for compat to be an issue, and this can be an `appearance: base`-only thing.

For filtering, I'd expect _both_ textContent _and_ the label (if provided) to be used.

**Comment by josepharhar** [comment]
From the meeting, it sounds like there are different use cases for text representations of options:
- what the text is rendered when the option is in the button
- what the text is rendered when the option is in the picker
- what text is used when doing typeahead (non-filterable select)
- what text is used to match for filtering in a filterable select

**Comment by lukewarlow** [comment]
This example is where I think using label might be problematic. It replaces the contents of the option, I believe this was for consistency with appearance: auto. Changing this to render the contents but then use label for the search/filter/typeahead would be problematic, I don't think we can change stuff like slot assignment based on the used appearance value.

`data:text/html,<select><option label="Hello">F</option></select><style>select, ::picker(select) { appearance: base-select; }</style>`

<img width="113" height="78" alt="Image" src="https://github.com/user-attachments/assets/7a5a3853-393d-431b-847f-79341b8d0d76" />

**Comment by sb3nder** [comment]
@lukewarlow
(I don't have experience with `label`) What does a screen reader announce when `label` is used?

**Comment by lukewarlow** [comment]
Tested on Chrome using VoiceOver and it reads the label as if it was the contents. So as far as I can tell it replaces the options contents for everything that the user interacts with (except for the form submission which uses the contents as the value).

**Comment by LeaVerou** [comment]
> Tested on Chrome using VoiceOver and it reads the label as if it was the contents. So as far as I can tell it replaces the options contents for everything that the user interacts with (except for the form submission which uses the contents as the value).

I think that’s a plus, completely consistent with the framing that label is the same content, but in (possibly shorter) text-only format. Since a screen-reader is inherently text-only, and benefits from brevity.

If we want to _actually_ replace the contents with label in certain cases (e.g. when selected) we can always make sure this works:

```css
option[label] {
	content: attr(label);
}
```

**Comment by sb3nder** [comment]
I also think that not rendering `label` would be nice, it makes the fallback story much easier.

It would also be slightly more consistent with `<optgroup>`, which doesn't render the `label` in base appearance when there is a `<legend>`, while in auto appearance it always renders the `label`.

```html
<select>
  <optgroup label="opt-label">
    <legend>opt-legend</legend>
    <option>Velociraptor</option>
  </optgroup>
</select>
```

<img width="222" height="183" alt="Image" src="https://github.com/user-attachments/assets/5b8402b8-d6b0-4d0e-9a78-487cd532c6d5" />

<img width="225" height="159" alt="Image" src="https://github.com/user-attachments/assets/e4f4c670-3ac5-44a4-b0c8-ecd3fe7682e5" />

**Comment by sb3nder** [comment]
> It would also be slightly more consistent with <optgroup>, which doesn't render the label in base appearance when there is a <legend>, while in auto appearance it always renders the label.

Actually after testing, in chrome for the `<optgroup>` the label attribute take precedence over the `<legend>` only when the picker is rendered with the platform ui.
In firefox it is complicated, in webkit from what I seen the `<legend>` take precedence.

Test: https://codepen.io/sb3nder/full/vExgzbV
